src/tigerbeetle/client.zig

```diff
@@ -0,0 +1,734 @@
+const std = @import("std");
+const builtin = @import("builtin");
+const assert = std.debug.assert;
+
+const constants = @import("../constants.zig");
+const stdx = @import("../stdx.zig");
+
+const build_options = @import("vsr_options");
+
+const vsr = @import("vsr");
+const IO = vsr.io.IO;
+const MessagePool = vsr.message_pool.MessagePool;
+
+const tb = vsr.tigerbeetle;
+
+pub fn ClientType(comptime StateMachine: type, comptime MessageBus: type) type {
+    return struct {
+        const Self = @This();
+        const Client = vsr.Client(StateMachine, MessageBus);
+
+        fn err(comptime m: []const u8, args: anytype) noreturn {
+            const stderr = std.io.getStdErr().writer();
+            stderr.print(m, args) catch unreachable;
+            std.os.exit(1);
+        }
+
+        fn print(comptime m: []const u8, args: anytype) void {
+            const stdout = std.io.getStdOut().writer();
+            stdout.print(m, args) catch return;
+        }
+
+        pub const Context = struct {
+            event_loop_done: bool,
+            request_done: bool,
```

Comment:
I'd suggest you follow something similar to [`tick_until_state_change`]( 
https://github.com/tigerbeetledb/tigerbeetle/blob/965978139836969f7d9bc9e629026d41b4e68be5/src/lsm/tree_fuzz.zig#L188-L193).

It's a great pattern to convert the async callbacks into a single-threaded program like the CLI, and allows asserting if we are in the right state when submitting/receiving messages.

The idea is to deal only with synchronous functions inside the loop.

```zig
while (!context.event_loop_done) {
    // ...
    switch (command) {
        .create_accounts => {
            const accounts = context.create_accounts(body) catch |err| {
                print_err(err);
                continue;
            };
            print_accounts(accounts);
        },
    // ...
    }
}
```

Reply:
I don't understand what you mean, since that (to me) is basically what is being done at line 509:

```
509
510            while (!context.event_loop_done) {
511                if (context.request_done and context.repl) {
512                    // Gets reset after every execution.
513                    var execution_arena = &std.heap.ArenaAllocator.init(arena.allocator());
514                    defer execution_arena.deinit();
515                    repl(context, execution_arena) catch return;
516                }
517                context.client.tick();
518                try io.run_for_ns(constants.tick_ms * std.time.ns_per_ms);
519            }
```